### DISCUSSION TITLE
Losing the biggest developer and having to fork hurt; let's do better to work together

### DISCUSSION THREAD

**Discussion by glasswings** [FossifyOrg/General-Discussion#39] [General]
Linux is immune to corporate buyouts. [Here's why](https://github.com/torvalds/linux/graphs/contributors?from=2002-04-07&to=2023-12-06&type=a) - look at how many different people have contributed large amounts of code.

[The same logic applies to Blender](https://github.com/blender/blender/graphs/contributors?from=2002-10-06&to=2023-12-06&type=a).  There are too many people who would need to be bought out.

[Contrast Simple Gallery](https://github.com/SimpleMobileTools/Simple-Gallery/graphs/contributors?from=2016-02-14&to=2023-12-06&type=a), the flagship app before the fork.  A reasonably likely downside scenario for Zipoapps was that they might get a C&D and have to remove a few hundred lines of translations here and there.  But as long as they got Tibbi's permission they were 99% clear.  (And, like, don't blame Tibbi for selling.  The option existed because there wasn't a high volume of community code.)

So, how does a project transition from a single large contributor to many smaller contributors?  The self-evident answer is we can't just make Naveen do it.  Even if you financially support the project (please do!) hiring one person to do everything doesn't make it a many-person effort.

I would like to suggest a bazaar-style of development, which means anyone can identify problems and suggest solutions - Naveen's essential role would be to listen, evaluate, and make things official.  Other kinds of work can and should be done by anyone who finds motivation and makes the time.  These three tasks seem particularly important to me:

---

### Eliminate new-developer friction

A technically curious user should be able to go from "hey what if the app did XYZ?" to "my phone is running a development build" as easily as possible.  The jump from user to developer is already difficult enough if people need to learn Kotlin and Android APIs.  Ideally the effort required to set up tools should be as little as possible.

Take Wikipedia as an example.  Even if you don't know how to write in encyclopedic style you can still click the "Edit" tab and holy cow there's the markdown.  "Install the app from source code" should be a task that someone can do even without knowing Kotlin or Android development.  Playing with an existing open source app may be their first real programming project.

I'll try setting up a dev environment.  I'm pretty weak in Kotlin and haven't done Android dev - the fact that I don't know actually helps me give a realistic assessment of how difficult it is to set up.  More perspectives are needed.

---

### Support users and identify pain-points

The SimpleMobileTools repository could go dark at any moment and take hundreds of issues down - those are hundreds of issues for people to work on.  Someone should archive those - [maybe try this software](https://github.com/topics/github-issues-export).

(To be polite to GitHub's servers we don't need to scrape the issues 200 times.  Maybe, like, 6 is reasonable.  Comment if you work on this.)

We need to replace that infrastructure, at the very least breathe life into the Fossify issue trackers.  If you think you can troubleshoot or try to reproduce bugs, go ahead and watch the projects now.  If there are too many issues you can unwatch or change settings later.

User spaces on Discord or Matrix (Matrix is cool but most users will be more familiar with Discord) seem like a good idea too.  If you set something up, remember that you're not official but your moderation choices do have an impact on the community's reputation.

---

### Work on issues if you can contribute anything; you don't have to code a solution

Fixing a bug is like a relay race.  If you know how to code and test you're able to run those last segments, but there are other important steps, especially "[Document steps to reproduce an issue](https://www.drupal.org/community/contributor-guide/task/document-steps-to-reproduce-an-issue)" (Drupal Contributor Guide).  Often a user will identify something that frustrates them, but they don't know enough or have the time or energy to figure out specifically when it breaks.

An issue that caught my attention says that some Calendar years are missing Easter.  The reporter has given useful details - the years, some other holidays are missing, they're in Portugal.  This is a pretty good bug report but it can be better.  How can a developer sitting in Canada see the localized behavior for Portugal?  There has to be a setting in Android Studio...

Example: Find the setting.  Make a quick note where it is, so that a new developer or tester can get up to speed.  Verify issue in Portugal.  Try other locales that follow Roman Catholic holidays.  Try locales that don't.  Post a clear and concise follow-up.

[Drupal has a guide to their entire issue lifecycle](https://www.drupal.org/community/contributor-guide/reference-information/quick-info/life-cycle-of-an-issue).  It's not necessary to adopt the whole thing, but I think it's full of good ideas.

**Comment by gmasil** [comment]
I started to open PRs when I first found a bug in the Gallery app. Since then I was somewhat active ([my PRs](https://github.com/SimpleMobileTools/Simple-Gallery/pulls?q=is%3Apr+author%3Agmasil)).
I am concerned that one developer will be overwhelmed with all those apps/repos. PRs stayed open for quite some time, technical misunderstandings, language barriers etc.
Is there any plan on letting others join the team? I would certainly be interested in helping out with the Gallery app, maybe there could be some kind of trial period or whatever. Any plans in that direction?

**Reply by naveensingh** [reply]
> Is there any plan on letting others join the team? I would certainly be interested in helping out with the Gallery app, maybe there could be some kind of trial period or whatever. Any plans in that direction?

@gmasil no plans yet, check my comment [here](https://github.com/FossifyOrg/General-Discussion/issues/29#issuecomment-1841888565).

**Reply by inson1** [reply]
@gmasil hey just create PRs, try help people with questions, issues etc. and after some time it just comes, naveensingh would be crazy not accept more help, but yea its just hard to trust people, so they dont create more bugs etc. Giving somebody rights to commits, merging PRs is big show of trust

even without the rights, 10. PR is going to be easier and faster to be merged than the 1. PR

**Comment by glasswings** [comment]
Well, my experience with developer experience turned into "you do not have enough RAM for Android development."  Google says 8GB is enough.  It is not.

I subjected Android Studio to a 5.5 GB resident size limit by using systemd ([which is a thing you can try](https://utcc.utoronto.ca/~cks/space/blog/linux/SystemdForMemoryLimiting)) and that stopped thrashing from negatively affecting the rest of the system, but Android Studio still tends too OOM itself.  So I wish I could actively participate but it's just not happening with this computer.  Sorry

**Reply by naveensingh** [reply]
@glasswings [16GB ram sticks are relatively cheap now a days.](https://www.amazon.com/Crucial-16GB-Laptop-Memory-CT16G4SFRA32A/dp/B08C511GQH/ref=sr_1_5?keywords=laptop%2Bram&qid=1703262092&sr=8-5&th=1)

**Reply by naveensingh** [reply]
I have a 64GB but I find 16GB to be enough for most projects, I would suggest Sway or i3 wm if you are having a memory crisis. But 16GB will likely OOM if you open multiple projects or a single project + many browser tabs.

**Reply by Wintandre** [reply]
I get the same issue. I don't know if the problem is with the IDE specifically or gradle or the interaction between them. Sometimes I can run AS all day and it's fine. Other times, it will consume my 16GB in less than an hour - obviously firing up one or more AVDs is going to consume memory but sometimes switching to the layout designer and back again seems to affect things worse.  Likewise lots building/installing on AVDs.

Never got to the source of the issue and I simply restart AS when the memory creeps up high enough, a RAM monitor in a visible panel helps with this.

I'll have to look at that memory limiting. Running out of memory isn't handled very well by the system.

**Reply by inson1** [reply]
16GB is minimal amount needed imo, but its still quite low
even Google Chrome takes lot of Ram, 32GB ram is the new amount for decent computer

btw if you would debug on your phone it would help with the ram needed

**Reply by y1111** [reply]
some other considerations to compile in low RAM pcs . are you using the emulator in pc ? u can use a phone too conected by usb with adb . u can see and select the phone there in the "emulators"/ devices select box in the IDE .
u can also see the screen and manipulate the phone in pc using tiny simple library like scrcpy. link ( https://github.com/Genymobile/scrcpy )

about browsers open , check Thorium  https://thorium.rocks/  . its foss , look in git too . im using it(im using the avx2 one) and its taking way less ram then chrome. firefox also have more efficient forks with telemery removed like floorp and librewolf . 

and also the apps are quite small , maybe it can be edited and compiled in those online (without installation) IDEs like gitpod -even in free tiers .

**Comment by muxina-credo** [comment]
I agree with the message in thread. To prevent this project from dying out, we need more active developers.

**Comment by dp35654** [comment]
Since I don't know how to code and I'm older I'm not sure what you guys consider a bug? But I will try to explain a problem with Gallery and file's vs Amazon drive or photos are devices that Google loaded with security features that will try and sway you from downloading a 3rd party app they've made it so permission of media & file's are removed if you haven't used the app for a while which the phone doesn't know since it doesn't recognize the app like f-droid in the first place kept saying it wasn't installed, I've turned off the button allowing Google to remove permission for file's and Media and to delete any file's and the switch is back on it's infuriating that I have to spend hours going through the phone and finding all the ways Google can remove permission as it's in other places like other app's that use location for example, trying to update the app's is another challenge since the device turns off the app's ability to install unknown app's & updates are like a new download, I'm not sure if Amazon stay's open because of internet access? I believe they figured out a way around the devices supposed security setting's? What's amazing is I might not check Google file's for 6mth unless I get a notification to free up storage and the permissions are never removed! Main device Motorola Edge Android 13 Google owned Motorola back in 2012 not sure when they sold it to Leveno? Hopefully you guys can figure a work around also

**Comment by mattdale77** [comment]
I came across this project fairly recently ans assumed it to be fairly active. I would be very disappointed if it has stalled due to lack of developer interest. 

What is the current status? There are a few bugs that are affecting me but it looks like they have been open for quite some time without any action and I'm concerned I've switched into a dying project 😥

**Reply by naveensingh** [reply]
Updates are on the way.

The project isn't dying or anything, progress is slow because I also have other priorities and I work on Fossify in my free time (for now) which I've had less of lately. In contrast, SMT had a couple full-time developers (including me) with Tibor basically working 24x7.

**Reply by mattdale77** [reply]
> Updates are on the way.
> 
> The project isn't dying or anything, progress is slow because I also have other priorities and I work on Fossify in my free time (for now) which I've had less of lately. In contrast, SMT had a couple full-time developers (including me) with Tibor basically working 24x7.

Thanks for your response. I'll stick with it :)
